go/tools/go-upgrade/go_upgrade.go

```diff
@@ -0,0 +1,453 @@
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// currentGolangVersion gets the running version of Golang in Vitess
+// and returns it as a *version.Version.
+//
+// The file `./build.env` describes which version of Golang is expected by Vitess.
+// We use this file to detect the current Golang version of our codebase.
+// The file contains `goversion_min x.xx.xx`, we will grep `goversion_min` to finally find
+// the precise golang version we're using.
+func currentGolangVersion() (*version.Version, error) {
+	contentRaw, err := os.ReadFile("build.env")
+	if err != nil {
+		return nil, err
+	}
+	content := string(contentRaw)
+	idxBegin := strings.Index(content, "goversion_min") + len("goversion_min") + 1
+	idxFinish := strings.Index(content[idxBegin:], "||") + idxBegin
+	versionStr := strings.TrimSpace(content[idxBegin:idxFinish])
+	return version.NewVersion(versionStr)
+}
+
+func currentBootstrapVersion() (float64, error) {
+	contentRaw, err := os.ReadFile("Makefile")
+	if err != nil {
+		return 0, err
+	}
+	content := string(contentRaw)
+	idxBegin := strings.Index(content, "BOOTSTRAP_VERSION=") + len("BOOTSTRAP_VERSION=")
+	idxFinish := strings.IndexByte(content[idxBegin:], '\n') + idxBegin
+	versionStr := strings.TrimSpace(content[idxBegin:idxFinish])
+	f, err := strconv.ParseFloat(versionStr, 64)
+	if err != nil {
+		return 0, err
+	}
+	return f, nil
+}
+
+// getLatestStableGolangReleases fetches the latest stable releases of Golang from
+// the official website using the goDevAPI URL.
+// Once fetched, the releases are returned as version.Collection.
+func getLatestStableGolangReleases() (version.Collection, error) {
+	resp, err := http.Get(goDevAPI)
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+
+	body, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	var latestGoReleases []latestGolangRelease
+	err = json.Unmarshal(body, &latestGoReleases)
+	if err != nil {
+		return nil, err
+	}
+
+	var versions version.Collection
+	for _, release := range latestGoReleases {
+		if !release.Stable {
+			continue
+		}
+		if !strings.HasPrefix(release.Version, "go") {
+			return nil, fmt.Errorf("golang version malformatted: %s", release.Version)
+		}
+		newVersion, err := version.NewVersion(release.Version[2:])
+		if err != nil {
+			return nil, err
+		}
+		versions = append(versions, newVersion)
+	}
+	return versions, nil
+}
+
+// chooseNewVersion decides what will be the next version we're going to use in our codebase.
+// Given the current Golang version, the available latest versions and whether we allow major upgrade or not,
+// chooseNewVersion will return either the new version or nil if we cannot/don't need to upgrade.
+func chooseNewVersion(curVersion *version.Version, latestVersions version.Collection, allowMajorUpgrade bool) *version.Version {
+	selectedVersion := curVersion
+	for _, latestVersion := range latestVersions {
```

Comment:
We should first sort the slice to be safe, no? From my reading of this, I think it's assumed to be in ascending order?
```
sort.Sort(version.Collection(versions))
```

Reply:
The `latestVersions` slice is usually in descending order, the order is not affecting the behavior of the `chooseNewVersion` function.